aiida/engine/processes/functions.py

```diff
@@ -253,7 +257,18 @@ def _define(cls, spec):  # pylint: disable=unused-argument
                     else:
                         valid_type = (Data,)
 
-                    spec.input(arg, valid_type=valid_type, default=default, serializer=to_aiida_type)
+                    # If a default is defined and it is not a ``Data`` instance it should be serialized, but this should
+                    # be done lazily using a lambda, just as any port defaults should not define node instances directly
+                    # as is also checked by the ``spec.input`` call.
+                    if (
+                        default is not None and default != UNSPECIFIED and not isinstance(default, Data) and
+                        not callable(default)
+                    ):
+                        indirect_default = lambda value=default: to_aiida_type(value)
```

Comment:
I think aiida data type can apply to the Data and get the same type, which means we don't need `isinstance(default, Data)`? 
Suppose `value = orm.Int(5)` then `to_aiida_type(value)` should still work. If the `to_aiida_type` is designed to be should only apply to the raw python data type (which I think not, it should also support pass `orm.Str` to it and return the aiida type), then this `isinstance(default, Data)` can be put into `to_aiida_type` and raise an exception.

Also, is that `callable(default)` would rule out the possible to serialize the python function to AiiDA type if it get supported in the future?

Reply:
> I think aiida data type can apply to the Data and get the same type, which means we don't need `isinstance(default, Data)`?

Nope, that raises:
```ipython
In [1]: from aiida.orm import to_aiida_type

In [2]: to_aiida_type(Int(2))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-d0b2d1ab4290> in <module>
----> 1 to_aiida_type(Int(2))

/usr/lib/python3.9/functools.py in wrapper(*args, **kw)
    886                             '1 positional argument')
    887 
--> 888         return dispatch(args[0].__class__)(*args, **kw)
    889 
    890     funcname = getattr(func, '__name__', 'singledispatch function')

~/code/aiida/env/dev/aiida-core/aiida/orm/nodes/data/base.py in to_aiida_type(value)
     19 def to_aiida_type(value):
     20     """Turns basic Python types (str, int, float, bool) into the corresponding AiiDA types."""
---> 21     raise TypeError(f'Cannot convert value of type {type(value)} to AiiDA type.')
     22 
     23 

TypeError: Cannot convert value of type <class 'aiida.orm.nodes.data.int.Int'> to AiiDA type.
```

Reply:
> Also, is that callable(default) would rule out the possible to serialize the python function to AiiDA type if it get supported in the future?

You mean that an arbitrary callable can get stored as a `Data` node? I guess you are right, but so far we haven't really had use cases where users want to pass arbitrary callables to process functions.

Now that you mention it, I have actually implemented such a `Data` plugin in `aiida-shell` for a different purpose (see this PR https://github.com/sphuber/aiida-shell/pull/19). It allows to store a function that is then called by the parser as a hook, allowing to provide custom parsing on the fly. For now it is used as an input to a `CalcJob` but I guess it could also be used as an input to a process function. I guess you are right that if we were to add this plugin to `aiida-core` then people could not rely on the automatic serialization of the default, but they can always do it manually, e.g.:
```python
from aiida_shell.data import PickledData
from aiida.engine import calcfunction

def my_callable():
    print('hello world')

@calcfunction
def test(func: callable = lambda: PickledData(my_callable)):
    func.load()()

test()  # will print 'hello world'

```

Reply:
Yes, it is just a nitpick from my side. If the user knows things like serializing the function they won't have a problem finding how it should be done of passing it.